Javascript Task Array-Problem

arbabkhanocloud opened the pull request:
In this array problem, for each array object's 'value' key, assign the next mentioned object to its 'value' property. Also, assign the next object to its nested object 'value' property and so on until the next object is null. 
Kindly check the file ( name: Array_Problem_Task_latest ).

arbabkhanocloud committed:
Javascript Task#3 ArrayProblem

In this array problem, for each array object, assign the next mentioned object to its 'value' property. Also, assign the next object to nested object 'value' property and so on until the next object is null.

arbabkhanocloud committed:
Create Array_Problem_Task_Recursion.js

In this method, I Implemented the array problem task with recursive function.

arbabkhanocloud commented on Array_Problem_Task_Recursion.js:
```diff
@@ -0,0 +1,48 @@
+});
+function arrrayProblemFUnction(arrayobject) { //recursive Function
+  if (arrayobject.next !== null) {
+    let check = findingNextObject(arrayobject.next);
```
##  findingNextObject( ) function is just checking and returning an object of given 'ID' if it exists in the array or does not exist. 
Code is complex because of this function only. As I did not use the arrays built-in function and implemented this function on my own.

bilalashraf1710 commented on Array_Problem_Task_Recursion.js:
```diff
@@ -0,0 +1,48 @@
+});
+function arrrayProblemFUnction(arrayobject) { //recursive Function
+  if (arrayobject.next !== null) {
+    let check = findingNextObject(arrayobject.next);
```
We should not use let if its not being reassigned somehwere.

arbabkhanocloud replied:
Noted.

bilalashraf1710 commented on Array_Problem_Task_Recursion.js:
```diff
@@ -0,0 +1,48 @@
+    arrrayProblemFUnction(arrayobject.value);
+    return arrayobject;
+  } else {
+    return arrayobject;
```
Never manipulate a parameter object, always create new object of it and return that.

arbabkhanocloud replied:
Noted.

bilalashraf1710 commented on Array_Problem_Task_Recursion.js:
```diff
@@ -0,0 +1,48 @@
+
+function findingNextObject(id) { // function to find next object in array
+  let index = 0;
+  var checkbool = false;
```
var is an old convention, use let or const accordingly.

arbabkhanocloud replied:
Noted.

bilalashraf1710 commented on Array_Problem_Task_Recursion.js:
```diff
@@ -0,0 +1,48 @@
+  var checkbool = false;
+  while (!checkbool && index < array2.length) {
+    if (array2[index].id === id) {
+      checkbool = true;
```
What are we trying to achieve with checkbool?

arbabkhanocloud replied:
checkbool is used to stop the while loop when the object is found in the array.

arbabkhanocloud committed:
Create Array_Problem_Task_latest.js

In this version, I have added a simplified  array problem task code and also changed the variable type from "var " to "const".

bilalashraf1710 commented on Array_Problem_Task_latest.js:
```diff
@@ -0,0 +1,27 @@
+  if (element.next === null) {
+    newArray.push(element);
+  } else {
+    let object = { ...element };
```
this should be const

arbabkhanocloud replied:
changed to "const"

bilalashraf1710 commented on Array_Problem_Task_latest.js:
```diff
@@ -0,0 +1,27 @@
+});
+// recursive function
+function recursiveFunction(arrayobject) {
+  if (arrayobject.next !== null) {
```
what if arrayobject is underfined?

arbabkhanocloud replied:
noted.

bilalashraf1710 commented on Array_Problem_Task_latest.js:
```diff
@@ -0,0 +1,27 @@
+    recursiveFunction(arrayobject.value);
+    return arrayobject;
+  } else {
+    return arrayobject;
```
we dont need this else statement, in both if and else conditions, we are just returning at the end, so we can just return outside the if else statements.

arbabkhanocloud replied:
yes, removed.

arbabkhanocloud committed:
Delete Array_Problem_Task_latest.js

arbabkhanocloud committed:
Create Array_Problem_Task_latest(2).js

In this commit, removed the else statement. changed object type "let" to "const" and check for "undefined" object

mahmad35 commented on Array_Problem_Task_Recursion.js:
```diff
@@ -0,0 +1,48 @@
+    if (array2[index].id === id) {
+      checkbool = true;
+      clone = { ...array2[index] };
+      index--;
```
When we are returning the element then why is there a need for decrementing index?

mahmad35 commented on Array_Problem_Task_Recursion.js:
```diff
@@ -0,0 +1,48 @@
+    index++;
+  }
+  if (checkbool === false) {
+    return false;
```
Since the function returns the object if found it should return null when the object is not found.

mahmad35 commented on Array_Problem_Task_latest(2).js:
```diff
@@ -0,0 +1,25 @@
+  if ((arrayobject !== undefined) && (arrayobject.next !== null)) {
+    const found = array.find((element) => element.id === arrayobject.next); //Finding the object in array of having 'id' equal to current object's 'next' value
+    const cloningObject = { ...found };
+    arrayobject.value = cloningObject; //assigning next object of array to current array_object's value
```
arrayObject.value = { ...found }

mahmad35 commented on JS_Task3_arrayProblem.js:
```diff
@@ -0,0 +1,29 @@
+                newArray.push(element);
+            }else{
+                    let tempObject = Object.assign(element)
+                    while( tempObject.next !== null){
```
can we reduce the time complexity to something like O(n^2)

arbabkhanocloud committed:
Update Array_Problem_Task_latest(2).js

Removed extra code of assigning found object to current array object.value

arbabkhanocloud committed:
Removed old file
